Prevent Query Rule Creation with Invalid Numeric Match Criteria

mridula-s109 opened the pull request:
This PR introduces validation logic for query rules that require numeric comparisons to prevent storing rules that will never match. Previously, invalid numeric criteria could be silently ignored at query time, leading to unexpected behavior.

- Added pre-validation in QueryRule.java to enforce numeric values for comparison rule types (lte, gte, lt, gt).
- If a rule contains invalid values, the API will return an explicit validation error instead of silently accepting the rule.
- Enhanced unit tests and YAML REST tests to verify validation errors.

mridula-s109 committed:
SEARCH-802 - bug fixed - Query rules allows for creation of rules with invalid match criteria

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -87,18 +86,77 @@ public static Map<String, Object> randomSearchApplicationQueryParams() {
+        List<Object> values = new ArrayList<>();
+        int numValues = randomIntBetween(1, 3);
+
+        if (criteriaType.isNumericComparison()) {
```
If you remove `isNumericComparison` you can directly check the types here rather than creating the methods to create numeric/nonNumeric criteria

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -119,4 +177,26 @@ public static Map<String, Object> randomMatchCriteria() {
+            }
+        } else {
+            // For other types (like BOOST), use a boost value
+            return Map.of("boost", randomDoubleBetween(0.0, 5.0, true));
```
Boost is not a valid query rule type, why was this added?

mridula-s109 replied:
my bad for other types ill just return an empty map

kderusso reviewed (commented):
Nice start @mridula-s109 ! I've left some feedback, please let me know if you have questions.

mridula-s109 committed:
Worked on the comments given in the PR

mridula-s109 commented:
> Nice start @mridula-s109 ! I've left some feedback, please let me know if you have questions.

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Fixed Integration tests

Commit:
[CI] Auto commit changes from spotless

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/10_query_ruleset_put.yml:
```diff
@@ -160,3 +170,51 @@ teardown:
+
+  - requires:
+      cluster_features: "query_rules.numeric_validation"
+      reason: "Query rules enforce numeric validation starting with 8.19"
```
Nitpick: 

```suggestion
      reason: "Fixed bug to enforce numeric rule validation starting with 8.19"
```

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/10_query_ruleset_put.yml:
```diff
@@ -160,3 +170,51 @@ teardown:
+      reason: "Query rules enforce numeric validation starting with 8.19"
+
+  - do:
+      catch: /\[query_ruleset\] failed to parse field \[rules\]/
```
Can we please catch the string pertaining to the error we're throwing here, to ensure that we're actually throwing the error that we think we are throwing? 

When you run a PUT command with an invalid ruleset, the response is: 
```
{
  "error": {
    "root_cause": [
      {
        "type": "x_content_parse_exception",
        "reason": "Failed to build [query_rule] after last required field arrived"
      }
    ],
    "type": "x_content_parse_exception",
    "reason": "Failed to build [query_rule] after last required field arrived",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "Input [bar] is not valid for CriteriaType [lte]"
    }
  },
  "status": 400
}
```

You can search on any part of that response, not just the top level parse error. You can see the reason at the bottom is an IllegalArgumentException with a more informative message: `Input [bar] is not valid for CriteriaType [lte]`

```suggestion
      catch: /Input \[not_a_number\] is not valid for CriteriaType \[lte\]/
```

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/10_query_ruleset_put.yml:
```diff
@@ -160,3 +170,51 @@ teardown:
+      reason: "Query rules enforce numeric validation starting with 8.19"
+
+  - do:
+      catch: /\[query_ruleset\] failed to parse field \[rules\]/
```
Same comment here RE: error message validation

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/10_query_ruleset_put.yml:
```diff
@@ -160,3 +170,51 @@ teardown:
+
+  - requires:
+      cluster_features: "query_rules.numeric_validation"
+      reason: "Query rules enforce numeric validation starting with 8.19"
```
Same nitpick here RE: reason

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeatures.java:
```diff
@@ -9,13 +9,20 @@
 import java.util.Set;
 
-public class EnterpriseSearchFeatures implements FeatureSpecification {
+public class EnterpriseSearchFeatures extends Plugin implements FeatureSpecification {
```
Why did you change this to extend `Plugin`?

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -86,19 +86,41 @@ public static Map<String, Object> randomSearchApplicationQueryParams() {
-
     public static QueryRule randomQueryRule() {
-        String id = randomIdentifier();
+        String ruleId = randomAlphaOfLengthBetween(3, 10);
```
It's OK to keep `randomIdentifier()` here.

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -53,6 +56,73 @@ public final void testRandomSerialization() throws IOException {
+                "ids": ["id1"]
+              }
+            }""");
+        QueryRule queryRule = QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON);
```
Why are we calling this rather than just passing in `testToXContentRules(content)`? (Question applies to all tests that do this)

mridula-s109 replied:
because in these cases an instance of randomquery rule is called, thus for each test case the inputs will be random.  Also i try to hardcode values as i am not a big fan of it.

kderusso replied:
I'm sorry, but I still don't understand. The content of the `testToXContentRules()` method is exactly identical to what you copied and pasted. I don't understand the need to duplicate it.

kderusso reviewed (commented):
Nice progress here @mridula-s109 , thanks for iterating! 

I've left a couple of very minor comments. 

The next thing that you'll want to do is: 

- Please update the title of this PR, to remove the Jira. It's OK to refer to the ticket number in the description (no links) but we don't want this being part of the official changelog as Jira is internal 🙂 
- Please remove this from draft state and apply the following labels:  
  - v9.0.0
  - v8.18.0
  - auto-backport
  - :SearchOrg/Relevance
  - `>bug`

This will automatically create a backport to the appropriate branch when it is merged, so the bug fix goes into 9.0. It will also create a Changelog entry for this bug fix. Finally, it will ping the Search organization that the PR is available. 

While CI failed, it looks like it failed on an unrelated change. CI will be re-run any time anything is added - so creating the changelog will run again. You may need to update from `main` via the button on the PR to resolve. We can't merge this PR until CI is green. 

Thanks for your work on this!

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

elasticsearchmachine commented:
Pinging @elastic/search-eng (Team:SearchOrg)

elasticsearchmachine commented:
Pinging @elastic/search-relevance (Team:Search - Relevance)

mridula-s109 committed:
Made changes from the PR

mridula-s109 committed:
Update docs/changelog/122823.yaml

elasticsearchmachine commented:
Hi @mridula-s109, I've created a changelog YAML for you.

Commit:
[CI] Auto commit changes from spotless

kderusso reviewed (commented):
One last question on tests. Please make sure that the breaking change proposal is approved before merging. Thank you!

kderusso committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Fixed the duplicate code issue in queryRuleTests

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -86,19 +86,41 @@ public static Map<String, Object> randomSearchApplicationQueryParams() {
+        return new QueryRule(ruleId, type, criteria, randomQueryRuleActions(), randomQueryRulePriority());
+    }
+
+    public static QueryRuleCriteria randomQueryRuleCriteria() {
+        QueryRuleCriteriaType criteriaType = randomFrom(QueryRuleCriteriaType.values());
+        String metadata = randomAlphaOfLengthBetween(3, 10);
+        List<Object> values = new ArrayList<>();
+        int numValues = randomIntBetween(1, 3);
+
+        if (List.of(QueryRuleCriteriaType.GT, QueryRuleCriteriaType.GTE, QueryRuleCriteriaType.LT, QueryRuleCriteriaType.LTE)
+            .contains(criteriaType)) {
+            for (int i = 0; i < numValues; i++) {
+                BigDecimal value = BigDecimal.valueOf(randomDoubleBetween(0, 1000, true)).setScale(2, RoundingMode.HALF_UP);
+                values.add(value.toString());
+            }
+        } else if (criteriaType == QueryRuleCriteriaType.ALWAYS) {
+            metadata = null;
+            values = null;
+        } else {
+            for (int i = 0; i < numValues; i++) {
+                values.add(randomAlphaOfLengthBetween(3, 10));
+            }
+        }
+
+        return new QueryRuleCriteria(criteriaType, metadata, values);
```
Nit: we could factor out the random value generation logic here into a `Function` and remove one of the `if` clauses

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -53,6 +56,66 @@ public final void testRandomSerialization() throws IOException {
+                "ids": ["id1"]
+              }
+            }""");
+        expectThrows(IllegalArgumentException.class, () -> QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON));
```
Can we test that the error message matches an expected value here, or at least contains some expected substring?

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -53,6 +56,66 @@ public final void testRandomSerialization() throws IOException {
+                "ids": ["id1"]
+              }
+            }""");
+        expectThrows(IllegalArgumentException.class, () -> QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON));
```
Same here

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -53,6 +56,66 @@ public final void testRandomSerialization() throws IOException {
+                "ids": ["id1"]
+              }
+            }""");
+        expectThrows(IllegalArgumentException.class, () -> QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON));
```
Same here

Mikep86 reviewed (commented):
LGTM, I left a couple minor comments. @kderusso is the query rules expert, so I will wait for her signoff :)

Mikep86 commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/55_numeric_validation.yml:
```diff
@@ -0,0 +1,138 @@
+  - match: { result: "created" }
+
+  - do:
+      catch: bad_request
```
@kderusso Is right here, this is an optimization. However, it can be valuable to use `catch: bad_request` to validate that the HTTP code is as expected. `catch: /your error message/` will not do that.

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Refactored code to clean it up based on PR comments

mridula-s109 commented:
> LGTM, I left a couple minor comments. @kderusso is the query rules expert, so I will wait for her signoff :)

Thanks Mike for your comments, I have actioned on them.

Commit:
[CI] Auto commit changes from spotless

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -99,27 +99,28 @@ public static QueryRule randomQueryRule() {
+        int numValues = randomIntBetween(1, 3);
+        List<Object> values = generateRandomValues(criteriaType, numValues);
         return new QueryRuleCriteria(criteriaType, metadata, values);
     }
```
Not really what I meant. I meant that we could take the logic that generates the random values:

```
                List.of(QueryRuleCriteriaType.GT, QueryRuleCriteriaType.GTE, QueryRuleCriteriaType.LT, QueryRuleCriteriaType.LTE)
                    .contains(criteriaType)
                        ? BigDecimal.valueOf(randomDoubleBetween(0, 1000, true)).setScale(2, RoundingMode.HALF_UP).toString()
                        : randomAlphaOfLengthBetween(3, 10)
```

and put that in an anonymous [Function](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html) that returns a String. There's no need to factor out this logic into a separate named function since it is only called in one place.

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -39,6 +41,7 @@
 
 public class QueryRuleTests extends ESTestCase {
     private NamedWriteableRegistry namedWriteableRegistry;
+    private static final Logger logger = LoggerFactory.getLogger(QueryRuleTests.class);
```
No need for a logger here, let's remove it

mridula-s109 replied:
i had removed it locally but looks like it was not published, ill do it now .

mridula-s109 replied:
I have made sure the changes are published, Thus resolving the conversation

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -83,7 +86,15 @@ public void testNumericValidationWithInvalidValues() throws IOException {
+            () -> QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON)
+        );
+        logger.info("Actual error message for invalid values: " + e.getMessage());
+        assertTrue(
+            "Error message [" + e.getMessage() + "] should contain 'Failed to build [query_rule]'",
+            e.getMessage().contains("Failed to build [query_rule]")
+        );
```
A couple things here:

- We tend to use `assertThat` and the Hamcrest matchers for these kind of checks (i.e. `assertThat(e.getMessage(), containsString("foo"))`
- We should check for the relevant part of the error message. In this case, that is "Input [abc] is not valid for CriteriaType [lte]"

Same goes for all the other error message checking assertions you added in this iteration.

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
resolved merge conflicts

# Conflicts:
#	x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Removed logger

mridula-s109 committed:
Trigger PR update

kderusso reviewed (commented):
Nice progress @mridula-s109 , I have some comments and requests before this is ready.

mridula-s109 committed:
Logger statements were removed

mridula-s109 committed:
Cleaned up the QueryRule tests

mridula-s109 commented:
> Nice progress @mridula-s109 , I have some comments and requests before this is ready.

I have resolved all the comments :)

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

kderusso reviewed (approved):
LGTM, thanks for the iterations!

Mikep86 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java:
```diff
@@ -86,19 +86,42 @@ public static Map<String, Object> randomSearchApplicationQueryParams() {
+        return new QueryRule(ruleId, type, criteria, randomQueryRuleActions(), randomQueryRulePriority());
+    }
+
+    private static List<Object> generateRandomValues(QueryRuleCriteriaType criteriaType, int numValues) {
+        if (criteriaType == QueryRuleCriteriaType.ALWAYS) {
+            return null;
+        }
+
+        List<Object> values = new ArrayList<>();
+        for (int i = 0; i < numValues; i++) {
+            values.add(
+                List.of(QueryRuleCriteriaType.GT, QueryRuleCriteriaType.GTE, QueryRuleCriteriaType.LT, QueryRuleCriteriaType.LTE)
+                    .contains(criteriaType)
+                        ? BigDecimal.valueOf(randomDoubleBetween(0, 1000, true)).setScale(2, RoundingMode.HALF_UP).toString()
+                        : randomAlphaOfLengthBetween(3, 10)
+            );
+        }
+        return values;
+    }
+
+    public static QueryRuleCriteria randomQueryRuleCriteria() {
+        QueryRuleCriteriaType criteriaType = randomFrom(QueryRuleCriteriaType.values());
+        String metadata = criteriaType == QueryRuleCriteriaType.ALWAYS ? null : randomAlphaOfLengthBetween(3, 10);
+        int numValues = randomIntBetween(1, 3);
+        List<Object> values = generateRandomValues(criteriaType, numValues);
+        return new QueryRuleCriteria(criteriaType, metadata, values);
     }
```
We could simplify this implementation into one method that is easier to read. Like this:

```suggestion
    public static QueryRuleCriteria randomQueryRuleCriteria() {
        Set<QueryRuleCriteriaType> numericValueCriteriaType = Set.of(QueryRuleCriteriaType.GT, QueryRuleCriteriaType.GTE, QueryRuleCriteriaType.LT, QueryRuleCriteriaType.LTE);
        QueryRuleCriteriaType criteriaType = randomFrom(QueryRuleCriteriaType.values());
        
        String metadata;
        List<String> values;
        if (criteriaType == QueryRuleCriteriaType.ALWAYS) {
            metadata = null;
            values = null;
        } else {
             metadata = randomAlphaOfLengthBetween(3, 10);
             values = new ArrayList<>();
             
             int numValues = randomIntBetween(1, 3);
             for (int i = 0; i < numValues; i++) {
                 values.add(
                     numericValueCriteriaType.contains(criteriaType)
                         ? BigDecimal.valueOf(randomDoubleBetween(0, 1000, true)).setScale(2, RoundingMode.HALF_UP).toString()
                         : randomAlphaOfLengthBetween(3, 10)
                 );
             }
        }

        return new QueryRuleCriteria(criteriaType, metadata, values);
    }
```

Mikep86 reviewed (commented):
Thanks for removing the logger! I have a suggestion about how to simplify the test code :)

mridula-s109 committed:
Update x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/EnterpriseSearchModuleTestUtils.java

Co-authored-by: Mike Pellegrini <mike.pellegrini@elastic.co>

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
Fixed checkstyle issues

mridula-s109 committed:
Merge branch 'main' into numeric-bugfix

mridula-s109 committed:
resolved compile errors

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 122823`

mridula-s109 commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/QueryRuleTests.java:
```diff
@@ -53,6 +57,79 @@ public final void testRandomSerialization() throws IOException {
+            () -> QueryRule.fromXContentBytes(new BytesArray(content), XContentType.JSON)
+        );
+        logger.info("Actual error message: " + e.getMessage());
+        assertTrue(e.getMessage().contains("failed to parse field [criteria]"));
```
@Mikep86 i am working on removing the logger and also will fix this nit as well in the followup PR.

mridula-s109 replied:
Have resolved it in this PR - https://github.com/elastic/elasticsearch/pull/124650

mridula-s109 commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124756">](https://github.com/elastic/elasticsearch/pull/124756)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

mridula-s109 commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124758">](https://github.com/elastic/elasticsearch/pull/124758)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
